apps/condo/domains/common/components/PostMessageProvider/Provider.tsx

```diff
@@ -0,0 +1,223 @@
+        version: { type: 'string', pattern: '^\\d+.\\d+.\\d+(?:-\\w+){0,1}$' },
+    },
+    required: ['handler', 'type', 'version', 'params'],
+    additionalProperties: false,
+}
+const validateMessage: ValidateFunction<MessageType> = ajv.compile(messageSchema)
+
+function getClientErrorMessage<Method extends AllRequestMethods> (
+    reason: ErrorReason,
+    message: string,
+    requestId?: RequestIdType,
+    method?: Method
+): ClientErrorResponse<AllRequestMethods, ErrorReason> {
+    return {
+        type: method ? `${method}Error` : 'CondoWebAppCommonError',
+        data: {
+            errorType: 'client',
+            errorCode: ERROR_CODES[reason],
+            errorReason: reason,
+            errorMessage: message,
+            ...(typeof requestId !== 'undefined' ? { requestId } : null),
+        },
+    }
+}
+
+/**
+ * Single place to handle all incoming postMessages from condo and its miniapps.
+ * You can register global handler for all origins by specifying "*" as origin (Example: notification)
+ * Or pass "per origin" handler (Example: window resize)
+ * Provider takes care of 3 things:
+ * 1. Making sure messages are trusted and coming from allowed Windows
+ * 2. Making sure message can be handled and passing parameters are valid
+ * 3. Handling errors thrown by individual handlers and converting it to corresponding postMessage response
+ */
+export const PostMessageProvider: React.FC = ({ children }) => {
+    const [allowedOrigins, setAllowedOrigins] = useState<Array<string>>([serverUrl])
+    const [registeredHandlers, setRegisteredHandlers] = useState<Record<HandlerOrigin, OriginHandlers>>({})
+    const isOnClient = typeof window !== 'undefined'
+
+    const registerHandler: RegisterHandler = useCallback((event, origin, handler) => {
+        setRegisteredHandlers(prev => {
+            return {
+                ...prev,
+                [origin]: {
+                    ...prev[origin],
+                    [event]: handler,
+                },
+            }
+        })
+    }, [])
+
+    const registerOrigin = useCallback((origin: string) => {
+        setAllowedOrigins((prev) => prev.includes(origin) ? prev : [...prev, origin])
+    }, [])
+
+    const resetOrigin = useCallback((origin: string) => {
+        setAllowedOrigins((prev) => prev.filter(element => element !== origin))
+        setRegisteredHandlers((prev) => omit(prev, origin))
+    }, [])
+
+    const handleMessage = useCallback((event: MessageEvent) => {
+        if (!event.isTrusted || !event.source ||
+            event.source instanceof ServiceWorker || event.source instanceof MessagePort) {
+            return
+        }
+
+        // If not valid message interface - skip. It can be system events, like webpack or some other legacy messages
+        if (validateMessage(event.data)) {
+            const message = event.data
+            const { requestId, ...params } = message.params
```

Comment:
Is the `requestId` guaranteed?
Maybe add an error or warning in the case of no `requestId`?
```suggestion
            const { requestId = null, ...params } = message.params
            
            if (!requestId) {
              // ...err
            }
```